File: data/mappers.py
from data.dtos import ResponseDTO
from domain.models.response_model import (
    FunctionCallResponseModel,
    ResponseModel,
    TextResponseModel,
)


class ResponseMapper:
    @staticmethod
    def to_domain(dto: ResponseDTO) -> ResponseModel:
        if dto.get_response_type() == "text":
            return TextResponseModel(
                content=dto.content,
                tokens_used=dto.tokens_used,
                model=dto.model,
                raw_response=dto.raw_response,
                finish_reason=dto.finish_reason,
            )
        elif dto.get_response_type() == "function_call":
            return FunctionCallResponseModel(
                function_name=dto.function_name,
                function_arguments=dto.function_arguments,
                tokens_used=dto.tokens_used,
                model=dto.model,
                raw_response=dto.raw_response,
                finish_reason=dto.finish_reason,
            )
        else:
            raise ValueError(
                f"Unknown DTO response type: {dto.get_response_type()}",
            )

File: src/data/mappers.ts
/**
 * Mappers to convert between DTOs and domain models.
 */

import type { ResponseDTO } from './dtos.js';
import { isTextResponseDTO, isFunctionCallResponseDTO } from './dtos.js';
import type { ResponseModel } from '../domain/models/response.js';

export class ResponseMapper {
  static toDomain(dto: ResponseDTO): ResponseModel {
    if (isTextResponseDTO(dto)) {
      return {
        type: 'text',
        content: dto.content,
        tokensUsed: dto.tokensUsed,
        model: dto.model,
        rawResponse: dto.rawResponse,
        finishReason: dto.finishReason,
      };
    } else if (isFunctionCallResponseDTO(dto)) {
      return {
        type: 'function_call',
        functionName: dto.functionName,
        functionArguments: dto.functionArguments,
        tokensUsed: dto.tokensUsed,
        model: dto.model,
        rawResponse: dto.rawResponse,
        finishReason: dto.finishReason,
      };
    } else {
      throw new Error('Unknown DTO response type');
    }
  }
}
